File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, val=0, next=None):
#     self.val = val
#     self.next = next
class Solution:
  def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    def merge2Lists(list1: Optional[ListNode], list2: Optional[ListNode]):
      dh = ListNode()
      curr1, curr2, currDH = list1, list2, dh
      while curr1 and curr2:
        if curr1.val <= curr2.val:
          currDH.next = curr1
          curr1 = curr1.next
        else:
          currDH.next = curr2
          curr2 = curr2.next
        currDH = currDH.next
      currDH.next = curr1 if curr1 else curr2
      return dh.next
    
    if len(lists) == 0: return None
    first = second = None
    while len(lists) > 1:
      first = lists.pop(0)
      second = lists.pop(0)
      lists.append(merge2Lists(first, second))
    return lists[0]

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func merge2Lists(list1 *ListNode, list2 *ListNode) *ListNode {
  dh, curr1, curr2 := ListNode{}, list1, list2
  currDH := &dh
  for curr1 != nil && curr2 != nil {
    if curr1.Val <= curr2.Val {
      currDH.Next = curr1
      curr1 = curr1.Next
    } else {
      currDH.Next = curr2
      curr2 = curr2.Next
    }
    currDH = currDH.Next
  }
  if curr1 != nil {
    currDH.Next = curr1
  } else {
    currDH.Next = curr2
  }
  return dh.Next
}

func mergeKLists(lists []*ListNode) *ListNode {
  if len(lists) == 0 {
    return nil
  }
  var first *ListNode
  var second *ListNode
  for len(lists) > 1 {
    first, second, lists = lists[0], lists[1], lists[2:]
    lists = append(lists, merge2Lists(first, second))
  }
  return lists[0]
}